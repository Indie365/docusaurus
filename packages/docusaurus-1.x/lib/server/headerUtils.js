#!/usr/bin/env node

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

const glob = require('glob');
const path = require('path');
const fs = require('fs-extra');
const metadataUtils = require('./metadataUtils.js');

// generate a doc header from metadata
function makeHeader(metadata) {
  let header = '---\n';
  Object.keys(metadata).forEach(key => {
    header += `${key}: ${metadata[key]}\n`;
  });
  header += '---\n';
  return header;
}

// rewriteHeaders rewrites all headers of markdown files under `filePath`
// changing `fromVersion` to `toVersion`
function rewriteHeaders(filePath, fromVersion, toVersion) {
  const files = glob.sync(`${filePath}/*`);
  files.forEach(file => {
    const stats = fs.lstatSync(file);
    if (stats.isDirectory()) {
      rewriteHeaders(file, fromVersion, toVersion);
      return;
    }
    const extension = path.extname(file);
    if (extension !== '.md' && extension !== '.markdown') {
      return;
    }
    const res = metadataUtils.extractMetadata(fs.readFileSync(file, 'utf8'));
    const metadata = res.metadata;
    const rawContent = res.rawContent;
    if (!metadata.id) {
      return;
    }
    metadata.id = metadata.id.replace(
      `version-${fromVersion}-`,
      `version-${toVersion}-`,
    );
    fs.writeFileSync(file, makeHeader(metadata) + rawContent);
  });
}

module.exports = {
  makeHeader,
  rewriteHeaders,
};
