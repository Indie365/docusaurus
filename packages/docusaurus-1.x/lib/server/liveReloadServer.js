/**
 * Copyright (c) 2017-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

const gaze = require('gaze');
const tinylr = require('tiny-lr');
const program = require('commander');
const readMetadata = require('./readMetadata.js');

function start(port) {
  process.env.NODE_ENV = 'development';
  process.env.LIVERELOAD_PORT = port;
  const server = tinylr();
  server.listen(port, () => {
    console.log('LiveReload server started on port %d', port);
  });

  gaze(
    [`../${readMetadata.getDocsPath()}/**/*`, '**/*', '!node_modules/**/*'],
    function() {
      this.on('all', () => {
        server.notifyClients(['/']);
      });
    },
  );
}

const getReloadScriptUrl = () => {
  // Please, JS programmer show me how to save state of 'program' globally for each
  // invocation of this function.
  
  // hacked in
  program
  .option('--port <number>', 'Specify port number')
  .option('--no-watch', 'Toggle live reload file watching')
  .option('--host <host>', 'use specified host (default: localhost)')
  .parse(process.argv);

  // hacked in
  program.parse(process.argv)

  const port = process.env.LIVERELOAD_PORT;
  const host = program.host || 'localhost';

  return `http://${host}:${port}/livereload.js`;
};

module.exports = {
  start,
  getReloadScriptUrl,
};
