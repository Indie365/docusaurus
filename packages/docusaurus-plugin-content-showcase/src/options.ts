/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {Joi, RouteBasePathSchema} from '@docusaurus/utils-validation';
import {GlobExcludeDefault} from '@docusaurus/utils';
import {tagSchema} from './tags';
import type {OptionValidationContext} from '@docusaurus/types';
import type {PluginOptions, Options} from '@docusaurus/plugin-content-showcase';

export const DEFAULT_OPTIONS: PluginOptions = {
  path: 'showcase', // Path to data on filesystem, relative to site dir.
  routeBasePath: '/showcase', // URL Route.
  include: ['**/*.{yml,yaml}'],
  // TODO exclude won't work if user pass a custom file name
  exclude: [...GlobExcludeDefault, 'tags.*'],
  screenshotApi: 'https://slorber-api-screenshot.netlify.app',
  tags: 'tags.yml',
};

const PluginOptionSchema = Joi.object<PluginOptions>({
  path: Joi.string().default(DEFAULT_OPTIONS.path),
  routeBasePath: RouteBasePathSchema.default(DEFAULT_OPTIONS.routeBasePath),
  include: Joi.array().items(Joi.string()).default(DEFAULT_OPTIONS.include),
  exclude: Joi.array().items(Joi.string()).default(DEFAULT_OPTIONS.exclude),
  tags: Joi.alternatives()
    .try(Joi.string().default(DEFAULT_OPTIONS.tags), tagSchema)
    .default(DEFAULT_OPTIONS.tags),
  screenshotApi: Joi.string().default(DEFAULT_OPTIONS.screenshotApi),
});

export function validateOptions({
  validate,
  options,
}: OptionValidationContext<Options, PluginOptions>): PluginOptions {
  const validatedOptions = validate(PluginOptionSchema, options);
  return validatedOptions;
}
