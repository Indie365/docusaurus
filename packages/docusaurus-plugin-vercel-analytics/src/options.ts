/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import {DEFAULT_PLUGIN_ID} from '@docusaurus/utils';
import {Joi} from '@docusaurus/utils-validation';
import type {OptionValidationContext} from '@docusaurus/types';

export type PluginOptions = {
  id: string;
  mode: 'auto' | 'production' | 'development' | undefined;
  debug: boolean | undefined;
};

export type Options = Partial<PluginOptions>;

const pluginOptionsSchema = Joi.object<PluginOptions>({
  id: Joi.string().valid(DEFAULT_PLUGIN_ID).default(DEFAULT_PLUGIN_ID),
  mode: Joi.string().valid('auto', 'production', 'development').optional(),
  debug: Joi.boolean().optional(),
});

export function validateOptions({
  validate,
  options,
}: OptionValidationContext<Options, PluginOptions>): PluginOptions {
  return validate(pluginOptionsSchema, options);
}
