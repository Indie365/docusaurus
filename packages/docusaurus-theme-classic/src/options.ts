/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {Joi} from '@docusaurus/utils-validation';
import type {
  ValidationResult,
  OptionValidationContext,
} from '@docusaurus/types';
import type {Options} from '@docusaurus/theme-classic';

const DEFAULT_OPTIONS: Options = {
  customCss: null,
};

export const Schema = Joi.object({
  customCss: Joi.alternatives()
    .try(Joi.array().items(Joi.string().required()), Joi.string().required())
    .optional()
    .default(DEFAULT_OPTIONS.customCss)
    .warning('deprecate.error', {
      msg: `theme.customCss option is deprecated.
Please use siteConfig.styling.css instead.

Note that this also changes the CSS insertion order!
This enables your site CSS to override default theme CSS more easily, without using !important

Before: custom site CSS was inserted before Infima CSS and default theme CSS.
After: custom site CSS will be inserted after Infima CSS and default theme CSS.

See also https://github.com/facebook/docusaurus/pull/6227
`,
    })
    .messages({
      'deprecate.error': '{#msg}',
    }),
});

export function validateOptions({
  validate,
  options,
}: OptionValidationContext<Options>): ValidationResult<Options> {
  return validate(Schema, options);
}
