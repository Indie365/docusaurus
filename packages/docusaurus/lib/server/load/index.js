/**
 * Copyright (c) 2017-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

const ejs = require('ejs');
const fs = require('fs-extra');
const _ = require('lodash');
const path = require('path');

const {generate} = require('@docusaurus/utils');
const loadConfig = require('./config');
const loadEnv = require('./env');
const loadTheme = require('./theme');
const loadRoutes = require('./routes');
const loadPlugins = require('./plugins');
const constants = require('../../constants');

module.exports = async function load(siteDir, cliOptions = {}) {
  const generatedFilesDir = path.resolve(
    siteDir,
    constants.GENERATED_FILES_DIR_NAME,
  );

  const siteConfig = loadConfig(siteDir);
  await generate(
    generatedFilesDir,
    constants.CONFIG_FILE_NAME,
    `export default ${JSON.stringify(siteConfig, null, 2)};`,
  );

  const env = loadEnv({siteDir, siteConfig});
  await generate(
    generatedFilesDir,
    'env.js',
    `export default ${JSON.stringify(env, null, 2)};`,
  );

  // Process plugins.
  const pluginConfigs = siteConfig.plugins || [];
  const context = {env, siteDir, generatedFilesDir, siteConfig, cliOptions};
  const {
    plugins,
    pluginsRouteConfigs,
    pluginsLoadedContent,
  } = await loadPlugins({
    pluginConfigs,
    context,
  });

  const outDir = path.resolve(siteDir, 'build');
  const {baseUrl} = siteConfig;

  // Resolve theme. TBD (Experimental)
  const themePath = loadTheme(siteDir);

  // Routing
  const {
    routesConfig,
    routesPaths,
    routesHashPath,
    routesHashPathFileName,
    routesMetadata,
    routesMetadataFileName,
    routesComponentStr,
    routesComponentFileName,
    routesAsyncModules,
    routesAsyncModulesFileName,
  } = await loadRoutes(pluginsRouteConfigs);

  await generate(
    generatedFilesDir,
    routesHashPathFileName,
    JSON.stringify(routesHashPath, null, 2),
  );
  await Promise.all(
    routesPaths.map(async routesPath => {
      const hashPath = routesHashPath[routesPath];
      const targetDir = path.join(
        generatedFilesDir,
        hashPath.replace(/^@generated\//, ''),
      );

      const metadata = routesMetadata[routesPath] || {};
      await generate(
        targetDir,
        routesMetadataFileName,
        JSON.stringify(metadata, null, 2),
      );

      const componentStr = routesComponentStr[routesPath];
      await generate(targetDir, routesComponentFileName, componentStr);

      const asyncModules = routesAsyncModules[routesPath] || [];
      await generate(
        targetDir,
        routesAsyncModulesFileName,
        JSON.stringify(asyncModules, null, 2),
      );
    }),
  );

  await generate(generatedFilesDir, 'routes.js', routesConfig);

  // -------------------------- TBD (Experimental) ----------------------
  // Generate contents metadata.
  const metadataTemplateFile = path.resolve(
    __dirname,
    '../../client/templates/metadata.template.ejs',
  );
  const metadataTemplate = fs.readFileSync(metadataTemplateFile).toString();
  const pluginMetadataImports = _.compact(pluginsLoadedContent).map(
    ({metadataKey, contentPath}) => ({
      name: metadataKey,
      path: contentPath,
    }),
  );

  const metadataFile = ejs.render(metadataTemplate, {
    imports: [
      ...pluginMetadataImports,
      {
        name: 'env',
        path: '@generated/env',
      },
    ],
  });
  await generate(generatedFilesDir, 'metadata.js', metadataFile);

  // ------------- END OF TBD -----------------------------------------

  const props = {
    siteConfig,
    siteDir,
    env,
    outDir,
    themePath,
    baseUrl,
    generatedFilesDir,
    routesPaths,
    plugins,
    cliOptions,
  };

  return props;
};
