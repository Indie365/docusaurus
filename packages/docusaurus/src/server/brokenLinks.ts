/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import _ from 'lodash';
import logger from '@docusaurus/logger';
import {matchRoutes} from 'react-router-config';
import {parseURLPath, serializeURLPath, type URLPath} from '@docusaurus/utils';
import {getAllFinalRoutes} from './utils';
import type {RouteConfig, ReportingSeverity} from '@docusaurus/types';

type BrokenLink = {
  link: string;
  resolvedLink: string;
  anchor: boolean;
};

type BrokenLinksByLocation = {[location: string]: BrokenLink[]};

type CollectedLinks = {
  [location: string]: {links: string[]; anchors: string[]};
};

function getPageBrokenLinks({
  allCollectedLinks,
  pagePath,
  pageLinks,
  routes,
}: {
  allCollectedLinks: CollectedLinks;
  pagePath: string;
  pageLinks: string[];
  pageAnchors: string[];
  routes: RouteConfig[];
}): BrokenLink[] {
  // console.log('routes:', routes);
  function isPathBrokenLink(linkPath: URLPath) {
    const matchedRoutes = [linkPath.pathname, decodeURI(linkPath.pathname)]
      // @ts-expect-error: React router types RouteConfig with an actual React
      // component, but we load route components with string paths.
      // We don't actually access component here, so it's fine.
      .map((l) => matchRoutes(routes, l))
      .flat();
    return matchedRoutes.length === 0;
  }

  function isAnchorBrokenLink(linkPath: URLPath) {
    const {pathname, hash} = linkPath;

    // Link has no hash: it can't be a broken anchor link
    if (hash === undefined) {
      return false;
    }

    const targetPage =
      allCollectedLinks[pathname] || allCollectedLinks[decodeURI(pathname)];

    // link with anchor to a page that does not exist (or did not collect any
    // link/anchor) is considered as a broken anchor
    if (!targetPage) {
      return true;
    }

    // it's a broken anchor if the target page exists
    // but the anchor does not exist on that page
    return !targetPage.anchors.includes(hash);
  }

  const brokenLinks = pageLinks.flatMap((link) => {
    const linkPath = parseURLPath(link, pagePath);
    if (isPathBrokenLink(linkPath)) {
      return [
        {
          link,
          resolvedLink: serializeURLPath(linkPath),
          anchor: false,
        },
      ];
    }
    if (isAnchorBrokenLink(linkPath)) {
      return [
        {
          link,
          resolvedLink: serializeURLPath(linkPath),
          anchor: true,
        },
      ];
    }
    return [];
  });

  return brokenLinks;
}

/**
 * The route defs can be recursive, and have a parent match-all route. We don't
 * want to match broken links like /docs/brokenLink against /docs/*. For this
 * reason, we only consider the "final routes" that do not have subroutes.
 * We also need to remove the match-all 404 route
 */
function filterIntermediateRoutes(routesInput: RouteConfig[]): RouteConfig[] {
  const routesWithout404 = routesInput.filter((route) => route.path !== '*');
  return getAllFinalRoutes(routesWithout404);
}

function getAllBrokenLinks({
  allCollectedLinks,
  routes,
}: {
  allCollectedLinks: CollectedLinks;
  routes: RouteConfig[];
}): {brokenLinks: BrokenLinksByLocation; brokenAnchors: BrokenLinksByLocation} {
  const filteredRoutes = filterIntermediateRoutes(routes);

  const allBrokenLinks = _.mapValues(
    allCollectedLinks,
    (pageCollectedData, pagePath) =>
      getPageBrokenLinks({
        allCollectedLinks,
        pageLinks: pageCollectedData.links,
        pageAnchors: pageCollectedData.anchors,
        pagePath,
        routes: filteredRoutes,
      }),
  );

  function splitBrokenLinks(collect: {[x: string]: BrokenLink[]}) {
    const brokenLinks: {[x: string]: BrokenLink[]} = {};
    const brokenAnchors: {[x: string]: BrokenLink[]} = {};

    Object.entries(collect).forEach(([page, links]) => {
      const [linksFiltered, anchorsFiltered] = _.partition(
        links,
        (link) => link.anchor === false,
      );

      if (linksFiltered.length > 0) {
        brokenLinks[page] = linksFiltered;
      }
      if (anchorsFiltered.length > 0) {
        brokenAnchors[page] = anchorsFiltered;
      }
    });

    return {brokenLinks, brokenAnchors};
  }

  return splitBrokenLinks(allBrokenLinks);
}

function brokenLinkMessage(brokenLink: BrokenLink): string {
  const showResolvedLink = brokenLink.link !== brokenLink.resolvedLink;
  return `${brokenLink.link}${
    showResolvedLink ? ` (resolved as: ${brokenLink.resolvedLink})` : ''
  }`;
}

function createBrokenLinksMessage(
  pagePath: string,
  allBrokenLinks: BrokenLink[],
): string {
  const type = allBrokenLinks[0]?.anchor === true ? 'anchor' : 'link';

  const anchorMessage =
    allBrokenLinks.length > 0
      ? `- Broken ${type} on source page path = ${pagePath}:
   -> linking to ${allBrokenLinks
     .map(brokenLinkMessage)
     .join('\n   -> linking to ')}`
      : '';

  return `${anchorMessage}`;
}

function getAnchorBrokenLinksErrorMessage(allBrokenLinks: {
  [location: string]: BrokenLink[];
}): string | undefined {
  if (Object.keys(allBrokenLinks).length === 0) {
    return undefined;
  }

  return `Docusaurus found broken anchors!

Please check the pages of your site in the list below, and make sure you don't reference any anchor that does not exist.
Note: it's possible to ignore broken anchors with the 'onBrokenAnchors' Docusaurus configuration, and let the build pass.

Exhaustive list of all broken anchors found:
${Object.entries(allBrokenLinks)
  .map(([pagePath, brokenLinks]) =>
    createBrokenLinksMessage(pagePath, brokenLinks),
  )
  .join('\n')}
`;
}

function getPathBrokenLinksErrorMessage(allBrokenLinks: {
  [location: string]: BrokenLink[];
}): string | undefined {
  if (Object.keys(allBrokenLinks).length === 0) {
    return undefined;
  }

  /**
   * If there's a broken link appearing very often, it is probably a broken link
   * on the layout. Add an additional message in such case to help user figure
   * this out. See https://github.com/facebook/docusaurus/issues/3567#issuecomment-706973805
   */
  function getLayoutBrokenLinksHelpMessage() {
    const flatList = Object.entries(allBrokenLinks).flatMap(
      ([pagePage, brokenLinks]) =>
        brokenLinks.map((brokenLink) => ({pagePage, brokenLink})),
    );

    const countedBrokenLinks = _.countBy(
      flatList,
      (item) => item.brokenLink.link,
    );

    const FrequencyThreshold = 5; // Is this a good value?
    const frequentLinks = Object.entries(countedBrokenLinks)
      .filter(([, count]) => count >= FrequencyThreshold)
      .map(([link]) => link);

    if (frequentLinks.length === 0) {
      return '';
    }

    return logger.interpolate`

It looks like some of the broken links we found appear in many pages of your site.
Maybe those broken links appear on all pages through your site layout?
We recommend that you check your theme configuration for such links (particularly, theme navbar and footer).
Frequent broken links are linking to:${frequentLinks}`;
  }

  return `Docusaurus found broken links!

Please check the pages of your site in the list below, and make sure you don't reference any path that does not exist.
Note: it's possible to ignore broken links with the 'onBrokenLinks' Docusaurus configuration, and let the build pass.${getLayoutBrokenLinksHelpMessage()}

Exhaustive list of all broken links found:
${Object.entries(allBrokenLinks)
  .map(([pagePath, brokenLinks]) =>
    createBrokenLinksMessage(pagePath, brokenLinks),
  )
  .join('\n')}
`;
}

export async function handleBrokenLinks({
  allCollectedLinks,
  onBrokenLinks,
  onBrokenAnchors,
  routes,
}: {
  allCollectedLinks: CollectedLinks;
  onBrokenLinks: ReportingSeverity;
  onBrokenAnchors: ReportingSeverity;
  routes: RouteConfig[];
}): Promise<void> {
  if (onBrokenLinks === 'ignore' && onBrokenAnchors === 'ignore') {
    return;
  }

  const {brokenLinks, brokenAnchors} = getAllBrokenLinks({
    allCollectedLinks,
    routes,
  });

  const pathErrorMessage = getPathBrokenLinksErrorMessage(brokenLinks);
  if (pathErrorMessage) {
    logger.report(onBrokenLinks)(pathErrorMessage);
  }

  const anchorErrorMessage = getAnchorBrokenLinksErrorMessage(brokenAnchors);
  if (anchorErrorMessage) {
    logger.report(onBrokenAnchors)(anchorErrorMessage);
  }
}
