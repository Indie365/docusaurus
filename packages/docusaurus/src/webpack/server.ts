/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import merge from 'webpack-merge';
import {
  NODE_MAJOR_VERSION,
  NODE_MINOR_VERSION,
  DOCUSAURUS_VERSION,
} from '@docusaurus/utils';
// Forked for Docusaurus: https://github.com/slorber/static-site-generator-webpack-plugin
import WebpackBar from 'webpackbar';
import {createBaseConfig} from './base';
import ssrDefaultTemplate from './templates/ssr.html.template';
import type {ServerEntryParams} from '../types';
import type {Props} from '@docusaurus/types';
import type {Configuration} from 'webpack';

type Params = Pick<
  ServerEntryParams,
  'onLinksCollected' | 'onHeadTagsCollected'
> & {
  props: Props;
};

function buildRoutesLocation({
  routesPaths,
  baseUrl,
}: {
  routesPaths: string[];
  baseUrl: string;
}) {
  const routesLocation: {[filePath: string]: string} = {};
  // Array of paths to be rendered. Relative to output directory
  routesPaths.forEach((str) => {
    const ssgPath =
      baseUrl === '/' ? str : str.replace(new RegExp(`^${baseUrl}`), '/');
    routesLocation[ssgPath] = str;
    return ssgPath;
  });
  return routesLocation;
}

export function createServerEntryParams(params: Params): ServerEntryParams {
  const {props, onLinksCollected, onHeadTagsCollected} = params;
  const {
    baseUrl,
    generatedFilesDir,
    headTags,
    preBodyTags,
    postBodyTags,
    outDir,
    siteConfig: {noIndex, ssrTemplate},
  } = props;

  const routesLocation: {[filePath: string]: string} =
    buildRoutesLocation(props);

  const manifestPath = path.join(generatedFilesDir, 'client-manifest.json');

  return {
    outDir,
    baseUrl,
    manifestPath,
    routesLocation,
    headTags,
    preBodyTags,
    postBodyTags,
    onLinksCollected,
    onHeadTagsCollected,
    ssrTemplate: ssrTemplate ?? ssrDefaultTemplate,
    noIndex,
    DOCUSAURUS_VERSION,
  };
}

export default async function createServerConfig(
  params: Params,
): Promise<Configuration> {
  const {props} = params;
  const config = await createBaseConfig(props, true);

  return merge(config, {
    target: `node${NODE_MAJOR_VERSION}.${NODE_MINOR_VERSION}`,
    entry: {
      main: path.resolve(__dirname, '../client/serverEntry.js'),
    },
    output: {
      filename: 'server.bundle.js',
      libraryTarget: 'commonjs2',
      // Workaround for Webpack 4 Bug (https://github.com/webpack/webpack/issues/6522)
      globalObject: 'this',
    },
    plugins: [
      // Show compilation progress bar.
      new WebpackBar({
        name: 'Server',
        color: 'yellow',
      }),
    ],
  });
}
