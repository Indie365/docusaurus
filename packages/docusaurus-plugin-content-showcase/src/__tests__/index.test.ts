/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import {loadContext} from '@docusaurus/core/src/server/site';
import {normalizePluginOptions} from '@docusaurus/utils-validation';
import {fromPartial} from '@total-typescript/shoehorn';
import pluginContentShowcase from '../index';
import {validateOptions} from '../options';
import type {PluginOptions} from '@docusaurus/plugin-content-showcase';

const loadPluginContent = async (siteDir: string, options: PluginOptions) => {
  const context = await loadContext({siteDir});
  const plugin = await pluginContentShowcase(
    context,
    validateOptions({
      validate: normalizePluginOptions,
      options,
    }),
  );
  return plugin.loadContent!();
};

describe('docusaurus-plugin-content-showcase', () => {
  it('loads simple showcase', async () => {
    const siteDir = path.join(__dirname, '__fixtures__', 'website');
    const showcaseMetadata = await loadPluginContent(
      siteDir,
      fromPartial({
        path: 'src/showcase',
        tags: 'tags.yaml',
      }),
    );

    expect(showcaseMetadata).toMatchSnapshot();
  });

  it('loads simple showcase with tags in options', async () => {
    const tags = {
      opensource: {
        label: 'Open-Source',
        description: {
          message:
            'Open-Source Docusaurus sites can be useful for inspiration!',
          id: 'showcase.tag.opensource.description',
        },
        color: '#39ca30',
      },
      meta: {
        label: 'Meta',
        description: {
          message: 'Docusaurus sites of Meta (formerly Facebook) projects',
          id: 'showcase.tag.meta.description',
        },
        color: '#4267b2',
      },
    };
    const siteDir = path.join(__dirname, '__fixtures__', 'website');
    const showcaseMetadata = await loadPluginContent(
      siteDir,
      fromPartial({
        path: 'src/showcase',
        tags,
      }),
    );

    expect(showcaseMetadata).toMatchSnapshot();
  });

  it('throw loading inexistant tags file', async () => {
    const siteDir = path.join(__dirname, '__fixtures__', 'website');
    await expect(async () => {
      await loadPluginContent(
        siteDir,
        fromPartial({
          path: 'src/showcase',
          tags: 'wrong.yaml',
        }),
      );
    }).rejects.toThrowErrorMatchingInlineSnapshot(
      `"Failed to read tags file for showcase"`,
    );
  });
});
