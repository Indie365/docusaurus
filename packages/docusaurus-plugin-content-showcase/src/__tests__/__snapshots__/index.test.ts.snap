// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`docusaurus-plugin-content-showcase loads simple showcase 1`] = `
{
  "items": [
    {
      "description": "World",
      "preview": "github.com/ozakione.png",
      "source": "https://github.com/facebook/docusaurus",
      "tags": [
        "opensource",
        "meta",
      ],
      "title": "Hello",
      "website": "https://docusaurus.io/",
    },
  ],
}
`;

exports[`docusaurus-plugin-content-showcase loads simple showcase with tags in options 1`] = `
{
  "items": [
    {
      "description": "World",
      "preview": "github.com/ozakione.png",
      "source": "https://github.com/facebook/docusaurus",
      "tags": [
        "opensource",
        "meta",
      ],
      "title": "Hello",
      "website": "https://docusaurus.io/",
    },
  ],
}
`;
